File: java/linkedlist/DeleteAllOccurrences.java
 /**
 * Q16. Delete All Occurrences in Doubly Linked List
 * 
 * Problem:
 * Given the head of a doubly linked list and an integer target, delete all nodes in the linked list
 * with the value target and return the head of the modified linked list.
 * 
 * Example:
 * Input: head -> 1 <-> 2 <-> 3 <-> 1 <-> 4, target = 1
 * Output: head -> 2 <-> 3 <-> 4
 * Explanation: All nodes with the value 1 were removed.
 * 
 * Approach:
 * 1. Handle edge case: if head is null, return null
 * 2. Traverse the list and for each node:
 *    - If node value equals target:
 *      a. Update previous node's next pointer
 *      b. Update next node's previous pointer
 *      c. Update head if first node is deleted
 *    - Move to next node
 * 
 * Time Complexity: O(N) where N is the number of nodes
 * Space Complexity: O(1) as we only use pointers
 */

class ListNode {
    int val;
    ListNode next;
    ListNode prev;
    ListNode(int x) { val = x; }
}

public class DeleteAllOccurrences {
    public ListNode deleteAllOccurrences(ListNode head, int target) {
        // Handle edge case
        if (head == null) {
            return null;
        }

        ListNode current = head;
        
        // Traverse the list
        while (current != null) {
            if (current.val == target) {
                // Store next node before modifying pointers
                ListNode nextNode = current.next;
                
                // Update previous node's next pointer
                if (current.prev != null) {
                    current.prev.next = current.next;
                } else {
                    // If deleting first node, update head
                    head = current.next;
                }
                
                // Update next node's previous pointer
                if (current.next != null) {
                    current.next.prev = current.prev;
                }
                
                // Move to next node
                current = nextNode;
            } else {
                current = current.next;
            }
        }
        
        return head;
    }

    // Helper method to create a doubly linked list
    private static ListNode createDoublyLinkedList(int[] values) {
        if (values == null || values.length == 0) {
            return null;
        }

        ListNode head = new ListNode(values[0]);
        ListNode current = head;
        ListNode prev = null;

        for (int i = 1; i < values.length; i++) {
            current.next = new ListNode(values[i]);
            current.prev = prev;
            prev = current;
            current = current.next;
        }
        current.prev = prev;

        return head;
    }

    // Helper method to print the list
    private static void printList(ListNode head) {
        ListNode current = head;
        while (current != null) {
            System.out.print(current.val);
            if (current.next != null) {
                System.out.print(" <-> ");
            }
            current = current.next;
        }
        System.out.println();
    }

    // Main method for testing
    public static void main(String[] args) {
        DeleteAllOccurrences solution = new DeleteAllOccurrences();

        // Test case 1: Delete target from middle and end [1,2,3,1,4], target = 1
        int[] values1 = {1, 2, 3, 1, 4};
        ListNode head1 = createDoublyLinkedList(values1);
        System.out.println("Test case 1: Delete target from middle and end");
        System.out.println("Original list:");
        printList(head1);
        ListNode result1 = solution.deleteAllOccurrences(head1, 1);
        System.out.println("List after deleting 1:");
        printList(result1);

        // Test case 2: Delete target from start [1,1,2,3,4], target = 1
        int[] values2 = {1, 1, 2, 3, 4};
        ListNode head2 = createDoublyLinkedList(values2);
        System.out.println("\nTest case 2: Delete target from start");
        System.out.println("Original list:");
        printList(head2);
        ListNode result2 = solution.deleteAllOccurrences(head2, 1);
        System.out.println("List after deleting 1:");
        printList(result2);

        // Test case 3: Delete non-existent target [1,2,3,4], target = 5
        int[] values3 = {1, 2, 3, 4};
        ListNode head3 = createDoublyLinkedList(values3);
        System.out.println("\nTest case 3: Delete non-existent target");
        System.out.println("Original list:");
        printList(head3);
        ListNode result3 = solution.deleteAllOccurrences(head3, 5);
        System.out.println("List after deleting 5 (no change):");
        printList(result3);

        // Test case 4: Delete all nodes [1,1,1], target = 1
        int[] values4 = {1, 1, 1};
        ListNode head4 = createDoublyLinkedList(values4);
        System.out.println("\nTest case 4: Delete all nodes");
        System.out.println("Original list:");
        printList(head4);
        ListNode result4 = solution.deleteAllOccurrences(head4, 1);
        System.out.println("List after deleting all 1s:");
        printList(result4);

        // Test case 5: Empty list
        ListNode head5 = null;
        System.out.println("\nTest case 5: Empty list");
        ListNode result5 = solution.deleteAllOccurrences(head5, 1);
        System.out.println("Result of deleting from empty list: " + (result5 == null ? "null" : "not null"));
    }
}

File: javascript/linkedlist/DeleteAllOccurrences.js
/**
 * Delete All Occurrences
 * 
 * Problem:
 * Given the head of a linked list and a value, remove all nodes that have the given value.
 * 
 * Example:
 * Input: head = [1,2,6,3,4,5,6], val = 6
 * Output: [1,2,3,4,5]
 * Explanation:
 * - Remove all nodes with value 6
 * - Keep all other nodes in their original order
 * 
 * Approach:
 * 1. Use a dummy node to handle edge cases
 * 2. Use two pointers: prev and current
 * 3. While traversing:
 *    - If current node's value matches val, skip it
 *    - Otherwise, keep the node and move pointers
 * 4. Return the modified list
 * 
 * Time Complexity: O(N) where N is the number of nodes
 * Space Complexity: O(1) as we only use pointers
 */

class ListNode {
    constructor(val) {
        this.val = val;
        this.next = null;
    }
}

class DeleteAllOccurrences {
    /**
     * @param {ListNode} head
     * @param {number} val
     * @return {ListNode}
     */
    deleteAllOccurrences(head, val) {
        if (!head) return null;

        // Create dummy node
        const dummy = new ListNode(0);
        dummy.next = head;
        let prev = dummy;
        let current = head;

        while (current) {
            if (current.val === val) {
                // Skip the current node
                prev.next = current.next;
            } else {
                // Keep the current node
                prev = current;
            }
            current = current.next;
        }

        return dummy.next;
    }

    // Helper method to create a linked list
    static createList(values) {
        if (!values || values.length === 0) return null;
        
        const head = new ListNode(values[0]);
        let current = head;
        
        for (let i = 1; i < values.length; i++) {
            current.next = new ListNode(values[i]);
            current = current.next;
        }
        
        return head;
    }

    // Helper method to print the list
    static printList(head) {
        let current = head;
        const values = [];
        while (current) {
            values.push(current.val);
            current = current.next;
        }
        console.log(values.join(' -> '));
    }
}

// Test cases
function runTests() {
    const solution = new DeleteAllOccurrences();

    // Test case 1: Regular case
    console.log("Test case 1: Regular case");
    const list1 = DeleteAllOccurrences.createList([1, 2, 6, 3, 4, 5, 6]);
    console.log("Original list:");
    DeleteAllOccurrences.printList(list1);
    console.log("List after deleting 6:");
    DeleteAllOccurrences.printList(solution.deleteAllOccurrences(list1, 6));

    // Test case 2: Value at the start
    console.log("\nTest case 2: Value at the start");
    const list2 = DeleteAllOccurrences.createList([1, 1, 2, 3, 4]);
    console.log("Original list:");
    DeleteAllOccurrences.printList(list2);
    console.log("List after deleting 1:");
    DeleteAllOccurrences.printList(solution.deleteAllOccurrences(list2, 1));

    // Test case 3: Value at the end
    console.log("\nTest case 3: Value at the end");
    const list3 = DeleteAllOccurrences.createList([1, 2, 3, 4, 5, 5]);
    console.log("Original list:");
    DeleteAllOccurrences.printList(list3);
    console.log("List after deleting 5:");
    DeleteAllOccurrences.printList(solution.deleteAllOccurrences(list3, 5));

    // Test case 4: All nodes have the value
    console.log("\nTest case 4: All nodes have the value");
    const list4 = DeleteAllOccurrences.createList([2, 2, 2, 2]);
    console.log("Original list:");
    DeleteAllOccurrences.printList(list4);
    console.log("List after deleting 2:");
    DeleteAllOccurrences.printList(solution.deleteAllOccurrences(list4, 2));

    // Test case 5: Empty list
    console.log("\nTest case 5: Empty list");
    const list5 = null;
    console.log("Original list: null");
    console.log("List after deleting 1:");
    DeleteAllOccurrences.printList(solution.deleteAllOccurrences(list5, 1));

    // Test case 6: Single node list
    console.log("\nTest case 6: Single node list");
    const list6 = DeleteAllOccurrences.createList([1]);
    console.log("Original list:");
    DeleteAllOccurrences.printList(list6);
    console.log("List after deleting 1:");
    DeleteAllOccurrences.printList(solution.deleteAllOccurrences(list6, 1));

    // Test case 7: Value not in list
    console.log("\nTest case 7: Value not in list");
    const list7 = DeleteAllOccurrences.createList([1, 2, 3, 4, 5]);
    console.log("Original list:");
    DeleteAllOccurrences.printList(list7);
    console.log("List after deleting 6:");
    DeleteAllOccurrences.printList(solution.deleteAllOccurrences(list7, 6));
}

// Run the tests
runTests(); 